### PR TITLE
More docs for `DataFrame.partitions`

### LINKED ISSUE #9951: API for filtering dataframes by partition
I just came across a need for a way to filter a Dask DataFrame based on some partition-level characteristics. In this case it was a Dask Geopandas dataframe and I was checking whether the partition's spatial footprint overlaps with some shape (dask-geopandas adds a `spatial_extent` attribute to each Dask GeoDataFrame). In general, this feature would be helpful whenever you have a boolean mask *on the partitions*.

The solution was to use a mixture of `dd.concat` and `dd.get_partition`

Here's a simple example:

```python
>>> import dask.dataframe as dd
>>> df = dask.dataframe.demo.make_timeseries()
>>> mask = df.id.map_partitions(lambda p: len(p.unique()) > 260).compute()
>>> keep = dd.concat([df.get_partition(i) for i, m in enumerate(mask) if m])
>>> keep
Dask DataFrame Structure:
                 name     id        x        y
npartitions=6                                 
               object  int64  float64  float64
                  ...    ...      ...      ...
...               ...    ...      ...      ...
                  ...    ...      ...      ...
                  ...    ...      ...      ...
Dask Name: concat, 8 graph layers
```

(Note that that has 6 partitions, instead of the normal 11).

It's not too difficult of a workaround, but I think a moderate expansion API would be acceptable. It doesn't quite fit `get_partition`, since it's getting multiple partitions. Perhaps a `.filter_partitions` that takes a boolean sequence the same length as the number of partitions? Or a callable that's provided each partition and should return a boolean?

I think this will become especially useful after https://github.com/dask/dask/pull/9473, since users will have more partition-level properties to filter on *without* a compute (cc @rjzamora).

### PR DESCRIPTION
**Pull request by TomAugspurger**
Closes #9951

### CODE REVIEW

File: dask/dataframe/core.py [hunk 1]
```diff
@@ -1356,9 +1381,10 @@ def partitions(self):
         """Slice dataframe by partitions
 
         This allows partitionwise slicing of a Dask Dataframe.  You can perform normal
-        Numpy-style slicing but now rather than slice elements of the array you
+        Numpy-style slicing, including s but now rather than slice elements of the array you
```
Comment (ncclementi) [inline dask/dataframe/core.py:1384]: what is "s" where it says including "s"?

**Review by ncclementi — COMMENTED** [review]
This LGTM, had small comment that I think it might be a typo

File: docs/source/dataframe-indexing.rst [hunk 2]
```diff
@@ -143,3 +143,64 @@ Trying to select specific rows with ``iloc`` will raise an exception:
...
+
+Or even more complicated selections based on the data in the partitions
+themselves (at the cost of computing the DataFrame up until that point). For
+example, we can create a boolean mask with the partitions that have more than
+some number of unique IDs using :meth:`DataFrame.map_partitions`:
+
+.. code-block:: python
+
+   mask = ddf.id.map_partitions(lambda p: len(p.unique()) > 20).compute()
+   ddf.partitions[mask]
```
Comment (jrbourbeau) [inline docs/source/dataframe-indexing.rst:197]:
```suggestion
   >>> mask = ddf.id.map_partitions(lambda p: len(p.unique()) > 20).compute()
   >>> ddf.partitions[mask]
```

**Review by jrbourbeau — COMMENTED** [review]
Thanks for writing this up @TomAugspurger. Looks great overall, just left a couple of minor comments

**Comment by jrbourbeau** [comment]
Oh, also just noticed that doctests are failing

```
______________ [doctest] dask.dataframe.core._Frame.get_partition ______________
686         Returns
687         -------
688         Dask DataFrame or Series
689             The same type as the original object.
690 
691         Examples
692         --------
693         >>> import dask
694         >>> ddf = dask.datasets.timeseries(start="2021-01-01", end="2021-01-07", freq="1H")
695         >>> ddf.get_partition(0)
Differences (unified diff with -expected +actual):
    @@ -1,5 +1,5 @@
     Dask DataFrame Structure:
                      name     id        x        y
    -npartitions=1
    +npartitions=1                                 
     2021-01-01     object  int64  float64  float64
     2021-01-02        ...    ...      ...      ...

/home/runner/work/dask/dask/dask/dataframe/core.py:695: DocTestFailure
```

File: dask/dataframe/core.py (hunk 1 above)
Comment (TomAugspurger) [reply dask/dataframe/core.py]: Mistake on my part. Fixed.

File: dask/dataframe/core.py [hunk 3]
```diff
@@ -1224,9 +1252,6 @@ def random_split(self, frac, random_state=None, shuffle=False):
 
         >>> a, b, c = df.random_split([0.8, 0.1, 0.1], random_state=123)  # doctest: +SKIP
 
-        See Also
-        --------
-        dask.DataFrame.sample
```
[the remark this answers is not in the document]
Comment (TomAugspurger) [reply dask/dataframe/core.py]: I got lost in the file. Restored it :)

**Comment by TomAugspurger** [comment]
Dockets is fixed (the `npartitions=1` has a bunch of whitespace in the repr)

**Review by jrbourbeau — APPROVED** [review]
Thanks @TomAugspurger

**Pull request merged**